google/resource_storage_bucket_iampolicy.go

```diff
@@ -0,0 +1,122 @@
+package google
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	storagev1 "google.golang.org/api/storage/v1"
+)
+
+func resourceStorageBucketIAMPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceStorageBucketIAMPolicyCreate,
+		Read:   resourceStorageBucketIAMPolicyRead,
+		Update: resourceStorageBucketIAMPolicyUpdate,
+		Delete: resourceStorageBucketIAMPolicyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"policy_data": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: jsonPolicyDiffSuppress,
+				ValidateFunc:     validateV2IamPolicy,
+			},
+			"etag": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func setBucketIamPolicy(d *schema.ResourceData, config *Config) error {
+	bucket := d.Get("bucket").(string)
+	policy, err := unmarshalStorageIamPolicy(d.Get("policy_data").(string))
+	if err != nil {
+		return fmt.Errorf("'policy_data' is not valid for %s: %s", bucket, err)
+	}
+
+	_, err = config.clientStorage.Buckets.SetIamPolicy(bucket, policy).Do()
+	return err
+}
+
+func resourceStorageBucketIAMPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	if err := setBucketIamPolicy(d, config); err != nil {
```

Comment:
We usually like to do a `log.Printf("[DEBUG] Setting IAM policy for bucket %q", bucket)` before we make an API call, and a `log.Printf("[DEBUG] Set IAM policy for bucket %q", bucket)` after the request completes, to help debug.